firmware/phy_task.c

```diff
@@ -0,0 +1,800 @@
+    } while (init_start != 0x0);
+
+    // Override Init Complete and force low to Complete MRW
+    dfi_set_init_complete_ovr_reg_if(true, 0x1);
+
+    // Wait for PHY to deassert INIT_COMPLETE
+    do
+    {
+        init_complete = dfi_get_init_complete_status_reg_if();
+    } while (init_complete != 0x1);
+
+    // Set Post Work Done Override
+    fsw_ctrl_set_post_work_done(true, 0x0);
+
+    // Send MRW from DFI Buffer (This was filled during prep)
+    dfi_buffer_send_packets(&wddr.dfi.dfi_buffer, false);
+
+    // Must disable buffer when done
+    dfi_buffer_disable(&wddr.dfi.dfi_buffer);
+
+    /**
+     * Per DFI 5.0 Specification - Section 4.21:
+     *
+     * PHYUPDATE:
+     * If dfi_init_start and dfi_phyupd_req are both asserted, the MC is
+     * permitted to not assert dfi_phyupd_ack. In this case, the PHY should
+     * de-assert dfi_init_complete to initiate the frequency change protocol.
+     * In this case it must also de-assert the dfi_phyupd_req signal before
+     * asserting the dfi_init_complete signal again.
+     *
+     * PHYMASTER:
+     * If the dfi_init_start and dfi_phymstr_req signals are both asserted,
+     * the MC cannot assert the dfi_phymstr_ack signal. In this case,
+     * the PHY may de-assert the dfi_init_complete signal to initiate the
+     * frequency change protocol and it must also de-assert the dfi_phymstr_req
+     * signal before asserting the dfi_init_complete signal again.
+     */
+    if (fw_manager.fsm.dfi.currentState == &dfiPhyUpd)
+    {
+        // Abort PHYUPD
+        stateM_handleEvent(&fw_manager.fsm.dfi, &(struct event){FW_PHY_EVENT_PHYUPD_ABORT, NULL});
+        configASSERT(fw_manager.fsm.dfi.currentState == &dfiIdle);
+    }
+    else if (fw_manager.fsm.dfi.currentState == &dfiPhyMstrPending)
+    {
+        // Abort PHYMSTR
+        stateM_handleEvent(&fw_manager.fsm.dfi, &(struct event){FW_PHY_EVENT_PHYMSTR_ABORT, NULL});
+        configASSERT(fw_manager.fsm.dfi.currentState == &dfiIdle);
+    }
+
+    // Release Init Complete Override
+    dfi_set_init_complete_ovr_reg_if(false, 0x1);
+
+    // Allow PLL lock interrupt
+    pll_set_lock_interrupt_state(&wddr.pll, true);
+}
+
+/*-----------------------------------------------------------*/
+static void dfi_phymstr_pending_entry_handler(__UNUSED__ void *stateData,
+                                              struct event *event)
+{
+    // Get master signal states from event data
+    dfi_phymstr_req_t *req = (dfi_phymstr_req_t *) event->data;
+
+    // Request PHY MASTER control
+    dfi_phymstr_req(&wddr.dfi, req);
+
+    // Enable PHY MASTER ACK IRQ
+    enable_irq(MCU_FAST_IRQ_PHYMSTR_ACK);
+}
+
+/*-----------------------------------------------------------*/
+static void dfi_phymstr_abort(void *currentStateData,
+                             struct event *event,
+                             void *newStateData)
+{
+    disable_irq(MCU_FAST_IRQ_PHYMSTR_ACK);
+    dfi_phymstr_exit(&wddr.dfi);
+#if CONFIG_CAL_PERIODIC
```

Comment:
Just curious, will we always use a vComplete to tell the caller we are finished?  Or is this only because the periodic scheduler is in its own thread?

Reply:
This is only to coordinate with the Periodic Cal thread that exists within the firmware.